thermal-control-design.md

```diff
@@ -0,0 +1,261 @@
+
+Vendor specific init function and clean up function will be performed at the beginning of the task and at then end of the task.
+
+A thermal control daemon class will be deifined with above functions defined, vendors will be allowed to have their own implementation.
+
+![](https://github.com/keboliu/SONiC/blob/master/images/thermal-control.svg)
+
+### 3.2 Policy management
+
+Policies are defined in a json file for each hwsku, it's including:
+
+- Thermal control algorithm control, enabled this hwsku or not, fan speed value to set if not running;
+
+- FAN absence action, suspend the algorithm or not, fan speed value to set;
+
+- PSU absence action, suspend the algorithm or not, fan speed value to set.
+
+Below is an example for the policy configuration:
+
+	{
+	    "thermal_control_algorithm": {
+		"run_at_boot_up": true,
+		"fan_speed_when_suspend": 10000
+	    },
+	    "fan_absence": {
+		"action": {
+		    "thermal_control_algorithm": "disable",
+		    "fan_speed": 10000,
+		    "led_color": "red"
+		}
+	    },
+	    "psu_absence": {
+		"action": {
+		    "thermal_control_algorithm": "disable",
+		    "fan_speed": 10000,
+		    "led_color": "red"
+		}
+	    }
+	}
+
+In this configuration, thermal control algorithm will run on this device; in fan absence situation, the fan speed need to be set to 10000, the thermal control algorithm will be suspended and fan status led shall be set to red ; in psu absence situation, thermal control algorithm will be suspend, fan speed will be set to 10000 and psu status led shall be set to red.
+
+During daemon start, this configuration json file will be loaded and parsed, daemon will handle the thermal control algorithm run and fan speed set when predefined policy meet.
+
+## 4. CLI show command for temperature and fan design
+
+### 4.1 New CLI show command for temperature
+
+ adding a new sub command to the "show platform":
+ 
+	admin@sonic# show platform ?
+	Usage: show platform [OPTIONS] COMMAND [ARGS]...
+
+	  Show platform-specific hardware info
+
+	Options:
+	  -?, -h, --help  Show this message and exit.
+
+	Commands:
+	  mlnx         Mellanox platform specific configuration...
+	  psustatus    Show PSU status information
+	  summary      Show hardware platform information
+	  syseeprom    Show system EEPROM information
+	  temperature  Show fan status information
+	  
+out put of the new CLI
+
+	admin@sonic# show platform temperature
+	NAME    Temperature    High Threshold    Low Threshold    Critical High Threshold   Critical Low Threshold   Warning Status
+	-----  -------------  ----------------  ---------------  ------------------------- ------------------------ ----------------
+	CPU        85              110              -10                 120                        -20                    false
+	ASIC       75              100               0                  110                        -10                    false
+
```

Comment:
For platforms which does thermal logic via BMC ( Ex monitor high/log and take action accordingly) in these cases do we  need to display CLI's values or can we just show as N/A ?